balancer/rls/balancer.go

```diff
@@ -77,6 +79,42 @@ var (
 	clientConnUpdateHook = func() {}
 	dataCachePurgeHook   = func() {}
 	resetBackoffHook     = func() {}
+
+	cacheEntriesMetric = estats.RegisterInt64Gauge(estats.MetricDescriptor{
+		Name:        "grpc.lb.rls.cache_entries",
+		Description: "EXPERIMENTAL. Number of entries in the RLS cache.",
+		Unit:        "entry",
+		Labels:      []string{"grpc.target", "grpc.lb.rls.server_target", "grpc.lb.rls.instance_uuid"},
+		Default:     false,
+	})
+	cacheSizeMetric = estats.RegisterInt64Gauge(estats.MetricDescriptor{
+		Name:        "grpc.lb.rls.cache_size",
+		Description: "EXPERIMENTAL. The current size of the RLS cache.",
+		Unit:        "By",
+		Labels:      []string{"grpc.target", "grpc.lb.rls.server_target", "grpc.lb.rls.instance_uuid"},
+		Default:     false,
+	})
```

Comment:
Did you consider passing the following to the `dataCache` at creation time:
- `cacheEntriesMetric`
- `cacheSizeMetric`
- Value of the `grpc.target` label
- Value of the `grpc.lb.rls.instance_uuid` label

`grpc.lb.rls.server_target` label is not available at `Build` time (which is when the dataCache is created). So, we would have to add a method on the `dataCache` to update the value of this label.

Once the above is done, the `dataCache` can take care of updating the cache related metrics.
- `cacheEntriesMetric` is straight-forward
  - The only two operations that change the number of entries in the `dataCache` are `addEntry` and `deleteAndCleanup`. You could update this metric everytime you return from these methods of the `dataCache`.
- `cacheSizeMetric` is also straight-forward
  - There are three operations that change the current size of the `dataCache`
    - `addEntry`
    - `updateEntrySize`
    - `deleteAndCleanup`
  - And you could update the value of this metric when returning from these methods of the `dataCache`  

Let me know what you think about this approach.

Reply:
Discussed offline; decided to make these metrics be asynchronous polling gauges. So a callback will be passed from here to OpenTelemetry.

I grab the cache mu in the callback to read the caches current properties, and also need to read the most recent rls server target. I could do this through the helper you mentioned above but decided to just grab RLS state mu and read it from cfg and use that to record.

Reply:
Decided to do this and emit on synchronous gauges with cache mu held in cache operations. Just to note I think you only need to pass target at construction, the metrics variables are global to the balancer package, and since the cache is 1:1 with the balancer, I think I can just create a UUID in newDataCache (and record current cache size/entries is 0). I'll send this out in a new PR.

Reply:
Like your suggestion for picker, I also passed a metrics recorder.